pydantic/main.py

```diff
@@ -1094,18 +1095,33 @@ def _calculate_keys(self, *args: Any, **kwargs: Any) -> Any:
 RootModelRootType = typing.TypeVar('RootModelRootType')
 
 
+class _RootModelSpecialValues(enum.Enum):
+    no_value = enum.auto()
+
+
+_RootModelNoValue = _RootModelSpecialValues.no_value
+
+
 class RootModel(BaseModel, typing.Generic[RootModelRootType]):
     __pydantic_root_model__ = True
-    # TODO: Make `__pydantic_fields_set__` logic consistent with `BaseModel`, i.e. it should be `set()` if default value
-    # was used
-    __pydantic_fields_set__ = {'root'}  # It's fine having a set here as it will never change
     __pydantic_private__ = None
     __pydantic_extra__ = None
 
     root: RootModelRootType
 
-    def __init__(__pydantic_self__, root: RootModelRootType) -> None:  # type: ignore
+    def __init__(
+        __pydantic_self__, root: RootModelRootType | Literal[_RootModelNoValue] = _RootModelNoValue
+    ) -> None:  # type: ignore
         __tracebackhide__ = True
+        if root is not _RootModelNoValue:
+            _object_setattr(__pydantic_self__, '__pydantic_fields_set__', {'root'})
+        else:
+            _object_setattr(__pydantic_self__, '__pydantic_fields_set__', set())
+            root_field = __pydantic_self__.model_fields['root']
+            root = root_field.get_default(call_default_factory=True)
+            if not (__pydantic_self__.model_config.get('validate_default', False) or root_field.validate_default):
```

Comment:
Under what circumstances does this branch get hit? And does this match `BaseModel`?

Also, it seems to me we could do the `_object_setattr(__pydantic_self__, '__pydantic_fields_set__', set())` line in the else branch of this conditional, is that wrong? Maybe that's what @samuelcolvin was referring to below?

Reply:
This gets hit by default. This is the branch for "do not validate default" which is the default behavior.

We want `__pydantic_fields_set__ == set()` for all `Undefined` regardless of `validate_default`. Thus, it looks logical to just set it right away for the entire `Undefined` branch.

Reply:
The `__pydantic_fields_set__` is being set in `__new__` since the last PR update.